corona-warn-app/cwa-app-android
# Add debug log feature for non-tester builds (EXPOSUREAPP-4451)

This adds the option for users to record debug logs in production builds.
To help investigate issues, fix bugs, or even if you are just curious about the app.

New entry under: `Menu > App Information > Error report`
The entry opens a new screen with a short explanation and controls to start/stop a debug log and share it.

* Sharing is manual and can happen through the native "Share" dialog through an app of your choice that supports sharing .zip file (e.g. email)
* Controls:
    * Pressing "start" will plant a `DebugLogTree` that logs all statements in a buffered `Flow`
    * `DebugLogger` will subscribe to this and write each line to a file
    * Before each line is written there is the chance for modules of type `BugCensor` to anonymize certain values
    * Pressing "delete" will stop and delete the current recording. It's not possible to stop recording without deleting the logs.
    * While logging active you can share the current log file. It's zipped and exposed via URI from a specific subdirectory `cache/debuglog/shared`. Sharing again will delete the previously shared file. Stopping the recording will also delete the zip file.
* The implementation differs a bit from the rest of the app (injection/singletons) to allow for more useful debug logs
   * To be able to trigger recordings across app restarts and to be independent of other app logic, we save the desired recording state with a trigger file, and we check for it's existence
   * To be able to record bugs caused by injeciton, early in the apps lifecycle, `DebugLogger` is an `object` and setup before we setup injection. We need access to some injected classes though for allowing value censoring, so logging will be buffered for a short while and only written to disk once Dagger is initialized

### TODO

### Testing
* See package `de.rki.coronawarnapp.bugreporting.debuglog`
* Test app kills, device restarts
* Check logout on release builds
* Check log sharing
* Check log deletion on stop

<img src="https://user-images.githubusercontent.com/1439229/103626225-3160ed80-4f3c-11eb-91e8-64a1da6527fb.png" width="400"><img src="https://user-images.githubusercontent.com/1439229/103626228-31f98400-4f3c-11eb-99ef-af6e5b3d60a0.png" width="400">
